File: sentimental-credit/credit.py
# TODO
import re
import sys


def main():
    # Prompt for input
    number = input("Number: ")

    # Check for conditions compliance
    if re.fullmatch("3[47]\d{13}", number) and check(number):
        print("AMEX")
        sys.exit(0)
    elif re.fullmatch("5[1-5]\d{14}", number) and check(number):
        print("MASTERCARD")
        sys.exit(0)
    elif (re.fullmatch("4\d{12}", number) or re.fullmatch("4\d{15}", number)) and check(
        number
    ) == True:
        print("VISA")
        sys.exit(0)
    else:
        print("INVALID")


# Check for accomplishment of the Luhn algorithm
def check(number):
    # Append the digits of the number to the list, starting from the end
    digits = list()
    for digit in reversed(number):
        digits.append(int(digit))

    # Calculate the summ of digits using the Luhn algorithm
    sum = 0
    for i in range(1, len(digits), 2):
        digit = digits[i]
        digit *= 2
        if digit > 9:
            sum += (digit // 10) + (digit % 10)
        else:
            sum += digit
    for i in range(0, len(digits), 2):
        digit = digits[i]
        sum += digit

    # Return the value by the last digit of the summ
    if sum % 10 == 0:
        return True
    else:
        return False


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Prompt for input a number
    long number = get_long("Number: ");

    // Count length of the number
    int length = 0;
    for (long i = number; i > 0; i /= 10)
    {
        length++;
    }

    // Calculate checksum
    long divisor = 10;
    int first_digits = 0, summ = 0;
    for (int i = 0; i < length; i++)
    {
        // Calculate value of a digit
        int digit = (number % divisor) / (divisor / 10);

        // Increase the summ value
        if (i % 2 == 0)
        {
            summ += digit;
        }
        else
        {
            summ += ((digit * 2) % 10) + ((digit * 2) / 10);
        }

        // Calculate value of the first two digits
        if (i == digits_number - 3)
        {
            first_digits = number / divisor;
        }

        // Increase divisor
        divisor *= 10;
    }

    // Evaluate conditions
    if (length == 15 && (first_digits == 34 || first_digits == 37) && summ % 10 == 0)
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (first_digits > 50 && first_digits < 56) && summ % 10 == 0)
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && first_digits / 10 == 4 && summ % 10 == 0)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
